File: IRemember/app/src/main/java/ru/goodibunakov/iremember/dialog/EditTaskDialogFragment.java
package ru.goodibunakov.iremember.dialog;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TimePicker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.google.android.material.textfield.TextInputLayout;

import org.jetbrains.annotations.NotNull;

import java.util.Calendar;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import ru.goodibunakov.iremember.R;
import ru.goodibunakov.iremember.alarm.AlarmHelper;
import ru.goodibunakov.iremember.model.ModelTask;
import ru.goodibunakov.iremember.utils.Utils;

public class EditTaskDialogFragment extends DialogFragment {

    private EditingTaskListener editingTaskListener;
    private Unbinder unbinder;

    @BindView(R.id.dialogTaskTitle)
    TextInputLayout tilTitle;
    @BindView(R.id.dialogTaskDate)
    TextInputLayout tilDate;
    @BindView(R.id.dialogTaskTime)
    TextInputLayout tilTime;
    @BindView(R.id.date)
    EditText etDate;
    @BindView(R.id.title)
    EditText etTitle;
    @BindView(R.id.time)
    EditText time;
    @BindView(R.id.spinner_priority)
    Spinner spinner;

    public static EditTaskDialogFragment newInstance(ModelTask task) {
        EditTaskDialogFragment editTaskDialogFragment = new EditTaskDialogFragment();

        Bundle args = new Bundle();
        args.putString("title", task.getTitle());
        args.putLong("date", task.getDate());
        args.putInt("priority", task.getPriority());
        args.putLong("timestamp", task.getTimestamp());

        editTaskDialogFragment.setArguments(args);
        return editTaskDialogFragment;
    }

    @Override
    public void onAttach(@NotNull Context context) {
        super.onAttach(context);
        try {
            editingTaskListener = (EditingTaskListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implements EditingTaskListener");
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        Bundle args = getArguments();

        String title;
        long date, timestamp;
        int priority;
        ModelTask modelTask;

        if (args != null) {
            title = args.getString("title");
            date = args.getLong("date");
            priority = args.getInt("priority");
            timestamp = args.getLong("timestamp");
            modelTask = new ModelTask(title, date, priority, 0, timestamp);


            final AlertDialog.Builder builder = new AlertDialog.Builder(Objects.requireNonNull(getActivity()));

            LayoutInflater inflater = getActivity().getLayoutInflater();
            @SuppressLint("InflateParams") View container = inflater.inflate(R.layout.dialog_task, null);
            unbinder = ButterKnife.bind(this, container);

            builder.setTitle(R.string.editing_title);
            builder.setIcon(R.mipmap.ic_launcher);
            etTitle.setText(title);
            etTitle.setSelection(etTitle.length());
            if (date != 0) {
                etDate.setText(Utils.getDate(date));
                time.setText(Utils.getTime(timestamp));
            }

            tilTitle.setHint(getResources().getString(R.string.task_title));
            tilDate.setHint(getResources().getString(R.string.task_date));
            tilTime.setHint(getResources().getString(R.string.task_time));

            builder.setView(container);

            Calendar calendar = Calendar.getInstance();
            calendar.set(Calendar.HOUR_OF_DAY, calendar.get(Calendar.HOUR_OF_DAY) + 1);

            ArrayAdapter<String> priorityAdapter = new ArrayAdapter<>(getActivity(),
                    android.R.layout.simple_spinner_dropdown_item, getActivity().getResources().getStringArray(R.array.priority_array));

            spinner.setAdapter(priorityAdapter);
            spinner.setSelection(priority);
            spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    modelTask.setPriority(position);
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {
                    modelTask.setPriority(0);
                }
            });

            if (etDate.length() != 0 || time.length() != 0){
                calendar.setTimeInMillis(date);
            }

            etDate.setOnClickListener(v -> {
                if (etDate.length() == 0) {
                    etDate.setText("");
                }

                DatePickerController datePickerController = new DatePickerController() {
                    @Override
                    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                        calendar.set(Calendar.YEAR, year);
                        calendar.set(Calendar.MONTH, month);
                        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                        etDate.setText(Utils.getDate(calendar.getTimeInMillis()));
                    }
                };
                datePickerController.onCreateDialog(getActivity()).show();
            });

            time.setOnClickListener(v -> {

                if (time.length() == 0) {
                    time.setText("");
                }

                TimePickerController timePickerController = new TimePickerController() {
                    @Override
                    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
                        calendar.set(Calendar.MINUTE, minute);
                        calendar.set(Calendar.SECOND, 0);
                        time.setText(Utils.getTime(calendar.getTimeInMillis()));
                    }
                };
                timePickerController.onCreateDialog(getActivity()).show();
            });

            builder.setPositiveButton(R.string.dialog_ok, (dialog, which) -> {
                modelTask.setTitle(etTitle.getText().toString());
                if (etDate.length() != 0 || time.length() != 0) {
                    modelTask.setDate(calendar.getTimeInMillis());
                    AlarmHelper alarmHelper = AlarmHelper.getInstance();
                    alarmHelper.setAlarm(modelTask);
                }
                modelTask.setStatus(ModelTask.STATUS_CURRENT);
                editingTaskListener.onTaskEdited(modelTask);
                dialog.dismiss();
                Log.d("debug", "model = " + modelTask.toString(modelTask));
            });
            builder.setNegativeButton(R.string.dialog_cancel, (dialog, which) -> dialog.cancel());

            AlertDialog alertDialog = builder.create();

            alertDialog.setOnShowListener(dialog -> {
                Button positive = ((AlertDialog) dialog).getButton(DialogInterface.BUTTON_POSITIVE);
                if (etTitle.length() == 0) {
                    positive.setEnabled(false);
                    tilTitle.setError(getResources().getString(R.string.dialog_error_empty_title));
                }
                etTitle.addTextChangedListener(new TextWatcher() {
                    @Override
                    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                    }

                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {
                        if (s.length() == 0) {
                            positive.setEnabled(false);
                            tilTitle.setError(getResources().getString(R.string.dialog_error_empty_title));
                        } else {
                            positive.setEnabled(true);
                            tilTitle.setErrorEnabled(false);
                        }
                    }

                    @Override
                    public void afterTextChanged(Editable s) {

                    }
                });
            });

            return alertDialog;
        } else {
            return super.onCreateDialog(savedInstanceState);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    public interface EditingTaskListener {
        void onTaskEdited(ModelTask updatedTask);
    }
}

File: IRemember/app/src/main/java/ru/goodibunakov/iremember/presentation/view/dialog/EditTaskDialogFragment.kt
package ru.goodibunakov.iremember.presentation.view.dialog


import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import moxy.MvpAppCompatDialogFragment
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import ru.goodibunakov.iremember.R
import ru.goodibunakov.iremember.RememberApp
import ru.goodibunakov.iremember.databinding.DialogTaskBinding
import ru.goodibunakov.iremember.presentation.model.ModelTask
import ru.goodibunakov.iremember.presentation.presenter.EditTaskDialogPresenter
import ru.goodibunakov.iremember.presentation.utils.DateUtils
import ru.goodibunakov.iremember.presentation.utils.DateUtils.FORMAT_DATE_ONLY
import ru.goodibunakov.iremember.presentation.utils.DateUtils.FORMAT_TIME_ONLY
import ru.goodibunakov.iremember.presentation.view.dialog.AddingTaskDialogFragment.Companion.KEY_NEGATIVE
import ru.goodibunakov.iremember.presentation.view.dialog.AddingTaskDialogFragment.Companion.KEY_POSITIVE
import ru.goodibunakov.iremember.presentation.view.dialog.AddingTaskDialogFragment.Companion.REQUEST_CODE_DATE
import ru.goodibunakov.iremember.presentation.view.dialog.AddingTaskDialogFragment.Companion.REQUEST_CODE_TIME

class EditTaskDialogFragment : MvpAppCompatDialogFragment(), EditTaskDialogFragmentView {

    private var _binding: DialogTaskBinding? = null
    // This property is only valid between onCreateDialog and onDestroyView.
    private val binding get() = _binding!!

//    private lateinit var container: View
    private lateinit var positive: Button
    private lateinit var timePickerDialogFragment: TimePickerDialogFragment
    private lateinit var datePickerDialogFragment: DatePickerDialogFragment

    @InjectPresenter
    lateinit var editTaskDialogPresenter: EditTaskDialogPresenter

    @ProvidePresenter
    fun providePresenter(): EditTaskDialogPresenter {
        return EditTaskDialogPresenter(RememberApp.databaseRepository)
    }

    companion object {
        fun newInstance(task: ModelTask): EditTaskDialogFragment {
            val editTaskDialogFragment = EditTaskDialogFragment()

            val args = Bundle()
            args.putString("title", task.title)
            args.putLong("date", task.date)
            args.putInt("priority", task.priority)
            args.putLong("timestamp", task.timestamp)

            editTaskDialogFragment.arguments = args
            return editTaskDialogFragment
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        _binding = DialogTaskBinding.inflate(LayoutInflater.from(context))

        val args = arguments
        if (args != null) {
            val title = args.getString("title")!!
            val date = args.getLong("date")
            val priority = args.getInt("priority")
            val timestamp = args.getLong("timestamp")
            editTaskDialogPresenter.onCreateDialog(title, date, priority, timestamp)

            val builder = AlertDialog.Builder(activity as Context, R.style.AppThemeDialog)
//            container = View.inflate(context, R.layout.dialog_task, null)
            isCancelable = false

            builder.setTitle(R.string.editing_title)
            builder.setIcon(R.mipmap.ic_launcher)
            editTaskDialogPresenter.initTitle()

            if (date != 0L) {
                editTaskDialogPresenter.setDateToUI()
                editTaskDialogPresenter.setTimeToUI()
            }

            binding.dialogTaskTitle.hint = resources.getString(R.string.task_title)
            binding.dialogTaskDate.hint = resources.getString(R.string.task_date)
            binding.dialogTaskTime.hint = resources.getString(R.string.task_time)

            builder.setView(binding.root)

            builder.setPositiveButton(R.string.dialog_ok) { _, _ ->
                editTaskDialogPresenter.okClicked(binding.etTitle.text.toString().trim())

                if (binding.etDate.length() != 0 || binding.etTime.length() != 0) {
                    editTaskDialogPresenter.setDateToModel()
                }
                editTaskDialogPresenter.updateTask()
                editTaskDialogPresenter.dismissDialog()
            }
            builder.setNegativeButton(R.string.dialog_cancel) { _, _ -> editTaskDialogPresenter.cancelDialog() }


            val priorityAdapter = ArrayAdapter(
                requireActivity(),
                android.R.layout.simple_spinner_dropdown_item,
                requireActivity().resources.getStringArray(R.array.priority_array)
            )

            binding.spinnerPriority.adapter = priorityAdapter
            editTaskDialogPresenter.setPriorityToUI()

            binding.spinnerPriority.onItemSelectedListener =
                object : AdapterView.OnItemSelectedListener {
                    override fun onItemSelected(
                        parent: AdapterView<*>,
                        view: View?,
                        position: Int,
                        id: Long
                    ) {
                        editTaskDialogPresenter.itemSelected(position)
                    }

                    override fun onNothingSelected(parent: AdapterView<*>) {
                        editTaskDialogPresenter.nothingSelected()
                    }
                }

            val alertDialog = builder.create()

            alertDialog.setOnShowListener { dialog ->
                positive = (dialog as AlertDialog).getButton(DialogInterface.BUTTON_POSITIVE)

                if (binding.etTitle.length() == 0) {
                    editTaskDialogPresenter.titleEmpty()
                }

                binding.etTitle.addTextChangedListener(object : MyTextWatcher {
                    override fun onTextChanged(
                        s: CharSequence,
                        start: Int,
                        before: Int,
                        count: Int
                    ) {
                        editTaskDialogPresenter.onTextChanged(s)
                    }
                })

                binding.etTitle.isFocusableInTouchMode = true
                editTaskDialogPresenter.getTitleHasFocus()
                binding.etTitle.onFocusChangeListener =
                    View.OnFocusChangeListener { _, hasFocus ->
                        if (hasFocus) {
                            editTaskDialogPresenter.setTitleHasFocus(hasFocus)
                        }
                    }

                binding.etDate.setOnClickListener {
                    clearTitleFocus()
                    if (binding.etDate.length() == 0) {
                        editTaskDialogPresenter.setEmptyDateToEditText()
                    }
                    editTaskDialogPresenter.dateClicked()
                }

                binding.etTime.setOnClickListener {
                    if (binding.etTime.length() == 0) {
                        editTaskDialogPresenter.setEmptyTimeToEditText()
                    }
                    editTaskDialogPresenter.timeClicked()
                }
            }

            return alertDialog
        } else {
            return super.onCreateDialog(savedInstanceState)
        }
    }

    override fun setDateToUI(date: Long) {
        binding.etDate.setText(DateUtils.getDate(date, FORMAT_DATE_ONLY))
    }

    override fun setTimeToUI(time: Long) {
        binding.etTime.setText(DateUtils.getDate(time, FORMAT_TIME_ONLY))
    }

    override fun setEmptyDateToEditText() {
        binding.etDate.setText("")
    }

    override fun showDatePickerDialog() {
        datePickerDialogFragment = DatePickerDialogFragment()
        datePickerDialogFragment.setTargetFragment(this@EditTaskDialogFragment, REQUEST_CODE_DATE)
        datePickerDialogFragment.show(
            requireActivity().supportFragmentManager,
            "DatePickerDialogFragment"
        )
    }

    override fun showTimePickerDialog() {
        timePickerDialogFragment = TimePickerDialogFragment()
        timePickerDialogFragment.setTargetFragment(this@EditTaskDialogFragment, REQUEST_CODE_TIME)
        timePickerDialogFragment.show(
            requireActivity().supportFragmentManager,
            "TimePickerDialogFragment"
        )
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == REQUEST_CODE_TIME) {
                when (data?.extras?.getString(AddingTaskDialogFragment.TYPE)) {
                    KEY_POSITIVE -> {
                        getTimeFromIntent(data)
                        editTaskDialogPresenter.positiveTimeClicked()
                    }
                    KEY_NEGATIVE -> editTaskDialogPresenter.negativeTimeClicked()
                }
            } else if (requestCode == REQUEST_CODE_DATE) {
                when (data?.extras?.getString(AddingTaskDialogFragment.TYPE)) {
                    KEY_POSITIVE -> {
                        getDateFromIntent(data)
                        editTaskDialogPresenter.positiveDateClicked()
                    }
                    KEY_NEGATIVE -> editTaskDialogPresenter.negativeDateClicked()
                }
            }
        }
    }

    private fun getTimeFromIntent(data: Intent?) {
        val hourOfDay = data?.extras?.getInt(AddingTaskDialogFragment.HOUR)
        val minute = data?.extras?.getInt(AddingTaskDialogFragment.MINUTE)
        if (hourOfDay != null && minute != null) {
            editTaskDialogPresenter.timeSelected(hourOfDay, minute)
        }
    }

    private fun getDateFromIntent(data: Intent?) {
        val year = data?.extras?.getInt(AddingTaskDialogFragment.YEAR)
        val month = data?.extras?.getInt(AddingTaskDialogFragment.MONTH)
        val day = data?.extras?.getInt(AddingTaskDialogFragment.DAY)
        if (year != null && month != null && day != null) {
            editTaskDialogPresenter.dateSelected(year, month, day)
        }
    }

    override fun setTimeToEditText(time: String) {
        binding.etTime.setText(time)
    }

    override fun setDateToEditText(date: String) {
        binding.etDate.setText(date)
    }

    override fun setEmptyTimeToEditText() {
        binding.etTime.setText("")
    }

    override fun dismissDialog() {
        dialog?.dismiss()
    }

    override fun cancelDialog() {
        dialog?.cancel()
    }

    override fun setUIWhenTitleEmpty() {
        positive.isEnabled = false
        binding.dialogTaskTitle.error = resources.getString(R.string.dialog_error_empty_title)
    }

    override fun setUIWhenTitleNotEmpty() {
        positive.isEnabled = true
        binding.dialogTaskTitle.isErrorEnabled = false
    }

    override fun initTitle(title: String) {
        binding.etTitle.setText(title)
        binding.etTitle.setSelection(binding.etTitle.length())
    }

    override fun setTitleFocus(hasFocus: Boolean) {
        if (hasFocus) {
            binding.etTitle.requestFocus()
            dialog?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
            (requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).toggleSoftInput(
                InputMethodManager.SHOW_FORCED,
                InputMethodManager.HIDE_IMPLICIT_ONLY
            )
        }
    }

    override fun setPriorityToUI(priority: Int) {
        binding.spinnerPriority.setSelection(priority)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideSoftKeyboard()
        _binding = null
    }

    override fun closeTimeDialogFragment() {
        if (::timePickerDialogFragment.isInitialized) {
            timePickerDialogFragment.dismiss()
        }
    }

    override fun closeDateDialogFragment() {
        if (::datePickerDialogFragment.isInitialized) {
            datePickerDialogFragment.dismiss()
        }
    }

    private fun clearTitleFocus() {
        if (binding.etTitle.hasFocus()) {
            binding.etTitle.clearFocus()
            hideSoftKeyboard()
            editTaskDialogPresenter.setTitleHasFocus(false)
        }
    }

    private fun hideSoftKeyboard() {
        (requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(
            binding.etTitle.windowToken,
            0
        )
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityCreated(arg0: Bundle?) {
        super.onActivityCreated(arg0)
        dialog?.window?.attributes?.windowAnimations = R.style.DialogAnimation
    }
}